### DISCUSSION TITLE
Difference between different decoders like Broadway, Native, h264BSD and tinyh264

### DISCUSSION THREAD

**Discussion by sree143342** [NetrisTV/ws-scrcpy#30] [Q&A] [answered]
Can any one explain about what is the difference between different types of decoders.

**Comment by drauggres** [comment, marked as answer]
Updated 02.2022

* [mbebenita/Broadway](https://github.com/mbebenita/Broadway)
* ~[oneam/h264bsd](https://github.com/oneam/h264bsd)~ removed
* [udevbe/tinyh264](https://github.com/udevbe/tinyh264)
* [xevokk/h264-converter](https://github.com/xevokk/h264-converter) (used for "Native")
* [WebCodecs](https://w3c.github.io/webcodecs/)

First three have the same idea: decoder in C-code is complied to [WebAssembly](https://webassembly.org/), each frame is decoded and later drawn on [canvas](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API).
Also they support only baseline encoded video (see https://github.com/NetrisTV/ws-scrcpy/issues/25#issuecomment-792684702)

The native decoder uses [MSE](https://developer.mozilla.org/en-US/docs/Web/API/Media_Source_Extensions_API), i.e. it builds mp4 containers and feed them to [`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video). In theory it should be more efficient than other decoders, because your browser may use hardware acceleration to decode video, but in my tests it was not stable enough to be the only decoder.

WebCodecs: Decoding is done by browser built-in (software/hardware) media decoder. The result is drawn on canvas. At the moment, available only in [Chromium](https://www.chromestatus.com/feature/5669293909868544) and derivatives.


This project is prototype: research is one of its purposes, so I added in it all ways (all I could find) to decode h264-video stream.

**Comment by sree143342** [comment]
Thank  you for reply @drauggres .... You told first three have same idea but it may have minor difference between them. i am searching exact difference.

**Comment by drauggres** [comment]
All I can tell you is what I know from the front pages of these projects (I did not dig into their source code):
Broadway uses original decoder from android without modifications; h264bsd uses same code, but with some optimizations; tinyh264 is a fork of h264bsd with more modifications. 

In my experience Broadway is the most stable decoder of all included.

**Comment by sree143342** [comment]
Thank you for your information @drauggres
